gdal/alg/marching_squares/square.h

```diff
@@ -0,0 +1,425 @@
+private:
+
+    ValuedPoint center() const
+    {
+        return ValuedPoint( 
+                .5*(upperLeft.x + lowerRight.x),
+                .5*(upperLeft.y + lowerRight.y),
+                (  (isnan(lowerLeft.value) ? 0 : lowerLeft.value)
+                 + (isnan(upperLeft.value) ? 0 : upperLeft.value)
+                 + (isnan(lowerRight.value) ? 0 : lowerRight.value)
+                 + (isnan(upperRight.value) ? 0 : upperRight.value)
+                 ) / (4-nanCount));
+    }
+
+    ValuedPoint leftCenter() const
+    {
+        return ValuedPoint( 
+                upperLeft.x,
+                .5*(upperLeft.y + lowerLeft.y), 
+                isnan(upperLeft.value)
+                ?  lowerLeft.value
+                : (isnan(lowerLeft.value) ? upperLeft.value : .5*(upperLeft.value + lowerLeft.value)));
+    }
+
+    ValuedPoint lowerCenter() const
+    {
+        return ValuedPoint( 
+                .5*(lowerLeft.x + lowerRight.x),
+                lowerLeft.y, 
+                isnan(lowerRight.value)
+                ? lowerLeft.value
+                : (isnan(lowerLeft.value) ? lowerRight.value : .5*(lowerRight.value + lowerLeft.value)));
+    }
+
+    ValuedPoint rightCenter() const
+    {
+        return ValuedPoint( 
+                upperRight.x,
+                .5*(upperRight.y + lowerRight.y), 
+                isnan(lowerRight.value)
+                ? upperRight.value
+                : (isnan(upperRight.value) ? lowerRight.value : .5*(lowerRight.value + upperRight.value)));
+    }
+
+    ValuedPoint upperCenter() const
+    {
+        return ValuedPoint( 
+                .5*(upperLeft.x + upperRight.x),
+                upperLeft.y, 
+                isnan(upperLeft.value)
+                ? upperRight.value
+                : (isnan(upperRight.value) ? upperLeft.value : .5*(upperLeft.value + upperRight.value)));
+    }
+
+    unsigned char marchingCase(double level) const
+    {
+        return (level < fudge(level, upperLeft.value) ? UPPER_LEFT : ALL_LOW)
+            |  (level < fudge(level, lowerLeft.value) ? LOWER_LEFT : ALL_LOW)
+            |  (level < fudge(level, lowerRight.value) ? LOWER_RIGHT : ALL_LOW)
+            |  (level < fudge(level, upperRight.value) ? UPPER_RIGHT : ALL_LOW);
+
+    }
+
+    static double interpolate_(double level, double x1, double x2, double y1, double y2, bool need_split)
+    {
+        if (need_split)
+        {
+            // the two cases are here to avoid numerical roundup errors, for two points, we always compute
+            // the same interpolation.  this condition is ensured by the order left->right bottom->top in interpole calls
+            //
+            // to obtain the same value for border (split) and non-border element, we take the
+            // middle value and interpolate from this to the end
+            const double xm = .5*(x1 + x2);
+            const double ym = .5*(y1 + y2);
+            if ((fudge(level, y1) < level && level < fudge(level, ym)) 
```

Comment:
fudge(level, y1) and fudge(level, ym) could be assigned to variables as being used twice each in the condition